# Python Project - Crime data Exploratory Data Analysis


*Abstract*

The preprocessing phase of this Python project focuses on transforming raw crime data obtained from the UK Police data portal (data.police.uk) into a clean, structured format suitable for exploratory data analysis (EDA). The objective is to prepare the data for subsequent analysis and insights generation, particularly with regards to identifying desirable and undesirable locations for real estate sales.

This phase encompasses several key steps, including data collection, understanding, cleaning, integration with third-party datasets, and preprocessing. Through systematic data cleaning and preprocessing techniques, we aim to address issues such as missing values, outliers, and inconsistencies in the dataset. Additionally, we will integrate the crime data with relevant third-party datasets to enrich our analysis.

The preprocessing phase plays a critical role in ensuring the quality and reliability of the insights derived from the data. By establishing a solid foundation through effective preprocessing, we can uncover meaningful patterns and trends that will inform decision-making for stakeholders, including Nadine Green, Head of Sales.

This abstract provides an overview of the preprocessing phase, highlighting its significance in the overall project workflow and its role in enabling data-driven decision-making in the real estate domain.


## Table of Contents

1. [Introduction](#1.-Introduction)
2. [Project Planning Phase](#2.-Project-Planning-Phase)
    - 2.1 [Project Statement of Work](#2.1-Project-Statement-of-Work)
    - 2.2 [Trello Board](#2.2-Trello-Board)
3. [Execution Phase](#3.-Execution-Phase)
    - 3.1 [Jupyter Notebook with Preprocessing](#3.1-Jupyter-Notebook-with-Preprocivities)
    - 3.2 [Analysis Report](#3.2-Analysis-Report)
    - 3.3 [Stakeholder Management Presentation](#3.3-Stakeholder-Management-Presentation)
4. [Preprocessing Phase](#4.-Preprocessing-Phase)
    - 4.1 [Introduction](#4.1-Introduction)
    - 4.2 [Data Collection](#4.2-Data-Collection)
    - 4.3 [Data Understanding](#4.3-Data-Understanding)
    - 4.4 [Data Cleaning](#4.4-Data-Cleaning)
    - 4.5 [Data Integration](#4tion)
    - 4.6 .5-Data-Integra[Data Quality Assessment](#4.6-Data-Qual
    - 4.7 ent)ta-Preproce[Workflow Visualization](#4.7-Workflow-Visualization)
5. [Analysis Phase](#5.-Anal- 5.1 [Graph Creation](#graph-creation)
       - 5.1.1 [Exploring Crime Type Distribution](#exploring-crime-type-distribution)
       - 5.1.2 [Identifying Top Crimes](#identifying-top-crimes)
       - 5.1.3 [Visualizing Crime Trends Over Time](#visualizing-crime-trends-over-time) 
    -  5.2 [Deriving Insights](#deriving-insights)
       - 5.2.1 [Regional Analysis: Comparing Crime Rates](#regional-analysis-comparing-crime-rates)
       - 5.2.2 [Crime Mapping: Spatial Distribution Analysis](#crime-mapping-spatial-distribution-analysis)Deriving-Insights)
6. [Conclusion](#6.-Conclusion)


# 1. Introduction

Crime data analysis plays a crucial role in understanding the safety and security landscape of a region, which is of paramount importance for various stakeholders, including real estate companies. In this project, we embark on an exploratory data analysis (EDA) journey to delve into the crime data of the United Kingdom sourced from the UK Police data portal ([data.police.uk](https://data.police.uk/data/)).

As a part of a real estate compathe our objective is to generate actionable insights to assist in decision-making processes, particularly regarding property salThe Our stakeholder, Nadine Green, Head of Sales, is keen to gain insights into the desirability of locations and to identify any useful information that can aid in making informed decisions around sales strategies.

The scope of our analysis will focus on the three most expensive real estate areas outside of London in the last two (March 2022 to March 2024), namely Cambridge, Oxford, and Bristol. According to a detailed crime analysis on the top three most expensive towns outside of London ([reference](https://theweek.com/arts-life/property/953851/house-prices-most-expensive-towns-england-outside-london) - article published 11th March 2024), the average property prices for these areas are as follows:
- Cambridge: £468,100
- Oxford: £446,700
- Bristol: £335,900

The corresponding police forces for theseCambridgeshire Constabulary,  areas are Thames Val andley Police, Avon and Somerset Constabulary, repf two years.

The preprocessing phase will be meticulously executed to ensure that the data is clean, consistent, and structured in a manner conducive to meaningful analysis. This will involve tasks such as data collection, understanding, cleaning, integration with third-partyother  datasets, and prepd normalization.

Subsequently, we will proceed with the analysis phase, where we will create visualizations using tools s andch as Maaborn, or Tableau to present our findings. Deriving insights from these visualizations will be pivotal in providing valuable information to Nadine and other stakeholders.

Throughout the project, effective stakeholder management will be paramount. It is imperative to keep Nadine informed, engaged, and aligned with the project objectives, ensuring that the analysis aligns with her need
This projectconsists inly analyzing the data and presentintheur findingsaimingim to provide critical insights that will support data-driven decision-making in the real estate secto.
 real estate domain.
estate domain.
estate domain.
ights.



# 2. Project Planning Phase

The project planning phase is a crucial initial step that lays the foundation for successful project execution. This phase involves defining the project’s scope, objectives, and deliverables, as well as outlining the steps necessary to achieve these goals. Effective planning ensures that the project remains on track and aligned with stakeholder expectations. This phase includes the creation of a Project Statement of Work (SOW) and the setup of a Trello board for task management and tracking.

## 2.1 Project Statement of Work

The Project Statement of Work (SOW) is a formal document that defines the project's objectives, scope, deliverables, timelines, and key milestones. It serves as a comprehensive guide for the project team and stakeholders, ensuring that everyone has a clear understanding of what the project entails and what is expected. The SOW outlines the responsibilities of each team member and establishes the criteria for project success. By providing a detailed roadmap, the SOW helps to mitigate risks and manage stakeholder expectations effectively.

## 2.2 Trello Board

The Trello board is a vital tool for project management and collaboration. It provides a visual overview of the project tasks, timelines, and progress. Using Trello, we can organize tasks into lists and cards, assign responsibilities, set deadlines, and track the status of each task. This transparency ensures that all team members are aware of their responsibilities and deadlines, fostering accountability and efficient workflow. The Trello board also facilitates communication and coordination among team members, helping to ensure that the project progresses smoothly and stays on schedule.

The designed Trello board can be accesses through:[Trello board](https://trello.com/invite/b/3eHSa5h5/ATTI36ca268f4ac702eb1b7222fcaf04a01166B8B047/python-project-crime-data-exploratory-data-analysis)


# 3. Execution Phase

The execution phase of the project involves the actual implementation of the data preprocessing, exploratory data analysis (EDA), and stakeholder communication. This phase is critical as it translates the planning into actionable tasks and delivers the final analysis and insights. The execution will be carried out through two Jupyter Notebooks: one focused on preprocessing the data, and the other dedicated to EDA. Various Python packages such as Matplotlib, NumPy, Pandas for preprocessing, and Seaborn, Folium, and GeoPandas for analysis will be utilized. The results and insights will be documented in markdown cells within the Jupyter Notebooks to ensure clarity and ease of understanding. The phase will culminate in a detailed analysis report and a stakeholder management presentation.

## 3.1 Jupyter Notebook with Preprocessing

The preprocessing notebook will encompass all steps necessary to prepare the raw crime data for analysis. This includes data collection from the UK Police data portal, handling missing  and empty columnsv ensuring the data is correct bya, and integrating third-party datasets. Key Python packages to be used in this notebook include:
- **Pandas** for data manipulation
- **NumPy** for numerical operations
- **Matplotlib** for initial visualizations

These packages need to be imported at the beginning of the notebook:
```python
import pandas as pd
import numpy as np
import matplotlib```
This notebook will ensure that the data is clean, consistent, and ready for exploratory analysis.

## 3.2 Analysis Report

The analysis report will be an integral part of the Jupyter Notebooks, documented in markdown cells alongside the code. This report will detail the findings from the EDA, highlighting crime trends, patterns, and insights across the selected regions of Oxford, Cambridge, and Bristol. The analysis will include:

- General analysis between regions (crime rate)
- Distributions regarding crime type, outcomes, and top crimes
- Crime map visualizations using packages like Folium and GeoPandas

These packages need to be imported as follows:
```python
import seaborn as sns
import folium
import geopandas as gpd
```
This structured approach ensures that the analysis is both comprehensive and accessible to stakeholders.

## 3.3 Stakeholder Management Presentation

Effective communication with stakeholders, particularly Nadine Green, Head of Sales, is paramount. During a meeting on May 28th, we discussed the project scope and objectives, focusing on the analysis of crime data in the three most expensive real estate areas outside of London (Oxford, Cambridge, and Bristol) over the past two years. The presentation will include:

- Introduction and project overview
- Detailed breakdowns of crime types and occurrences in each area
- Visualizations of trends over time, designed for both technical and non-technical audiences
- Comprehensive documentation to aid stakeholder understanding and communication

The presentation, scheduled for May 30th between 4 and 5 PM, will be preceded by a follow-up email to Nadine on May 29th, ensuring all preparations are aligned with her expectations. The final deliverable will include both Jupyter Notebooks and a comprehensive analysis report to facilitate informed decision-making for real estate investments.
.pyplot as plt


# 4. Preprocessing Phase

The preprocessing phase is a critical step in the data analysis process, ensuring that raw data is transformed into a clean and structured format suitable for further analysis. This phase involves several key tasks, including data collection, understanding, cleaning, integration, and quality assessment. By meticulously preprocessing the data, we aim to establish a reliable foundation for subsequent exploratory data analysis (EDA) and insights generation.

## 4.1 Introduction

In this section, we introduce the preprocessing phase, outlining its significance and the key steps involved. We emphasize the importance of preparing the data thoroughly to ensure accurate and meaningful analysis. The introduction provides an overview of the tasks to be performed and sets the stage for the detailed processes that foBefore any data can be collected and preprocessed, the necessary libraries need to be imported into the jupiter notebook:ifying areas for improvement.


In [1]:
# Import necessary libraries
import pandas as pd #Data manipulation and analysis, including reading/writing data, cleaning, transformation, 
                    #and analysis of structured data with DataFrames and Series.
import numpy as np # Numerical computing, providing support for large multi-dimensional arrays and mathematical functions.
import os # Handling file paths, navigating the file system, and managing environment variables. used to get directory of the imported files

## 4.2 Data Collection

Data collection is the first step in the preprocessing phase. Here, we describe the sources of the datasets, including the UK Police data portal and any third-party datasets. We outline the steps taken to acquire the data and specify the chosen datasets from the portal, focusing on the three police forces: Thames Valley Police, Avon and Somerset Constabulary, and Cambridgeshire Constabulary.
Considering that the analysis wsa made on 3 police forces over 2 years with the data seperated by month in the website, a function was created to automate the reading of the files. The purpose of this function is to read multiple CSV files, each containing crime data for a specific month, and concatenate them into a single dataframe. This allows for easier analysis and manipulation of the data.

In [2]:

# define a function that reads the files according to the name and directory
def read_monthly_data(root_directory, start_year, start_month, end_year, end_month, postDateCSVFileName):
    """
    Read monthly CSV files and concatenate them into a single dataframe for a range of years and months.
    
    Args:
    - root_directory (str): Root directory containing the CSV files.
    - start_year (int): Start year for which to read the data.
    - start_month (int): Start month for which to read the data.
    - end_year (int): End year for which to read the data.
    - end_month (int): End month for which to read the data.
    - postDateCSVFileName (str): End of the file name that differs according to the police force.
    
    Returns:
    - df_combined (DataFrame): Combined dataframe containing data from all specified months.
    """
    dataframes = []  # Initialize an empty list to store the dataframes

    # It loops through the specified range of years and months, constructs the file path for each CSV file,
    # and attempts to read the file using pd.read_csv()
    for year in range(start_year, end_year + 1):
        for month in range(start_month, 13 if year < end_year else end_month + 1):
            try:
                # Construct the file path
                file_path = os.path.join(root_directory, f'{year}-{str(month).zfill(2)}', f'{year}-{str(month).zfill(2)}-{postDateCSVFileName}')
                
                # Read the CSV file
                df = pd.read_csv(file_path)
                
                # Append the dataframe to the list
                dataframes.append(df)
                
            except Exception as e:
                print(f"An error occurred while reading data for {year}-{month}: {e}")
                # If an error occurs during reading, it prints an error message but continues to the next file
    
    # Concatenate all dataframes into a single dataframe
    df_combined = pd.concat(dataframes, ignore_index=True)
    
    return df_combined

Overall, this function simplifies the process of reading and combining multiple CSV files into a single dataframe, making it easier to work with the crime data for analysis.
This block of code is aimed at combining data from multiple police forces into a single dataframe for further analysis. 

In [3]:
# Specifies the directory where the CSV files containing the crime data are located, this will need to be changed to be replicated in another computer
root_directory = r'C:\Users\MartaMoreira\Python_DataAnalytics\PoliceForceData' 
# Define the start date from which data will be collected
start_year = 2022
start_month = 3
# Define the end date up to which data will be collected
end_year = 2024
end_month = 3
# A list containing the file names for each police force's data
postDateCSVFileNames = ['thames-valley-street.csv', 'avon-and-somerset-street.csv', 'cambridgeshire-street.csv']

# Initialize an empty list to store dataframes from all sources
all_dataframes = []

# A loop iterates over each police force's file name in postDateCSVFileNames
for postDateCSVFileName in postDateCSVFileNames:
    'For each police force, the read_monthly_data() function is called to '
    'read the monthly CSV files and concatenate them into a single dataframe (combined_data_original)'
    combined_data_original = read_monthly_data(root_directory, start_year, start_month, end_year, end_month, postDateCSVFileName)
    all_dataframes.append(combined_data_original)

# Concatenate all the combined dataframes into a single dataframe
combined_data_final = pd.concat(all_dataframes, ignore_index=True)

## 4.3 Data Understanding

Understanding the data is crucial before proceeding with cleaning and analysis. In this section, we explore the structure of the datasets, providing summary statistics and initial observations. This step helps us identify any potential issues and gain insights into the data's characteristics.

### 4.3.1 Data inspection

In [4]:
# 1. Data inspection
combined_data_final.head() # Examine first few rows of the dataset

Crime ID    Month  \
0  7312a0dc35158e77ef2ae6eb6ee1ecc1c1946e20cb8065...  2022-03   
1  be5a15af709d612e9b39b444377be7b2b9afcca350194f...  2022-03   
2  09698779f4bae0791b25e70122738d8872c8e2cda88df0...  2022-03   
3  d7bb4e184301f37f0f8450b50f874e95a40d4d4e54edd1...  2022-03   
4  5734122a4b9e007d93087a44be315c18302b7d4ae17ac4...  2022-03   

            Reported by          Falls within  Longitude   Latitude  \
0  Thames Valley Police  Thames Valley Police  -0.972323  51.991460   
1  Thames Valley Police  Thames Valley Police  -0.972323  51.991460   
2  Thames Valley Police  Thames Valley Police  -0.970516  51.992128   
3  Thames Valley Police  Thames Valley Police  -0.970516  51.992128   
4  Thames Valley Police  Thames Valley Police  -0.996750  51.997191   

                 Location  LSOA code            LSOA name  \
0  On or near Osprey Walk  E01017648  Aylesbury Vale 001A   
1  On or near Osprey Walk  E01017648  Aylesbury Vale 001A   
2  On or near Lime Avenue  E01017648  Aylesbury Vale 001A   
3  On or near Lime Avenue  E01017648  Aylesbury Vale 001A   
4   On or near Westfields  E01017649  Aylesbury Vale 001B   

                     Crime type                       Last outcome category  \
0  Violence and sexual offences  Action to be taken by another organisation   
1  Violence and sexual offences  Action to be taken by another organisation   
2  Violence and sexual offences  Action to be taken by another organisation   
3  Violence and sexual offences  Action to be taken by another organisation   
4     Criminal damage and arson                 Unable to prosecute suspect   

   Context  
0      NaN  
1      NaN  
2      NaN  
3      NaN  
4      NaN

### 4.3.2 Data description

This section aims to understand the characteristics and distribution of the variables within our dataset. This exploration helps grasp the patterns, trends, and variability present in the data. Aiming to gain insights that will inform subsequent analysis and decision-making processes.

In [5]:
 #since in the dataset, the numerical variables are either latitude or longitude, the describe funciton won't be usefull to give us statistics
combined_data_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 814555 entries, 0 to 814554
Data columns (total 12 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   Crime ID               717466 non-null  object 
 1   Month                  814555 non-null  object 
 2   Reported by            814555 non-null  object 
 3   Falls within           814555 non-null  object 
 4   Longitude              781446 non-null  float64
 5   Latitude               781446 non-null  float64
 6   Location               814555 non-null  object 
 7   LSOA code              781445 non-null  object 
 8   LSOA name              781445 non-null  object 
 9   Crime type             814555 non-null  object 
 10  Last outcome category  717466 non-null  object 
 11  Context                0 non-null       float64
dtypes: float64(3), object(9)
memory usage: 74.6+ MB


The previous function shows that there are a total of 814,555 entries in the dataframe, each row corresponding to a specific incident of crime. In addition, the dataframe consists of 12 column, each with a different attribute of the data. There is a mix of data types: 3 columns are tyoe float64(numerical) and 9 columns are type object (text).
Some of the columns ('Crime ID', ' Longitude', 'Latitude', 'LSOA code', 'LSOA name' and 'Last outcome category') have null values that need to either be removed or normalised, this is shown in the difference between the total number of entries and the non-null count. 
In particular:
- 'Crime ID' - is a unique identifier for each crime incident. There are 717,466 non-null entries, indicating that some incidents may not have a Crime ID;
- 'Month' - Month in which the crime was reported;
- 'Reported by' - Organization or authority that reported the crime;
- 'Falls within' - Geographic area within which the crime falls. This column and the column 'Reported by' appear to be equal;
- 'Location' - Description of the location where the crime occurred;
- 'Longitude' and 'Latitude' - Geographic coordinates of the crime location;
- 'LSOA code' and 'LSOA name' - Lower Layer Super Output Area (LSOA) code and name associated with the crime location;
- 'Crime type' - Type or category of the crime;
- 'Last outcome category' - Outcome category of the crime investigation or resolution. There are 717,466 non-null entries;
- 'Context' - This column has all null values.

All the missing values, equal columns and empty columns need to be cleaned. In addition, checks on wheter the LSOA codes are correct and the latitude and longitude coordinates and within the cities in study need to be done.

## 4.4 Data Cleaning, Integration and Quality Assessment

Data cleaning addresses issues such as missing values, data type inconsistencies, and duplicate records. In this section, we detail the methods and techniques used to clean the data, ensuring that it is accurate and consistent. This step is essential for eliminating noise and preparing the data for integration and analysis.

Data integration involves merging the crime data with relevant third-party datasets to enrich our analysis. A dataset with a reference list was added and methods used to combine datasets and handle any discrepancies were used. This step enhances the dataset's value by providing additional context and information.

Assessing data quality is vital to ensure the reliability of our analysis. In this section, we evaluate the quality of the data, using specific metrics to identify any remaining issues or concerns. In this specific scenario, we conducted several checks to ensure data accuracy. These included verifying if the Latitude and Longitude coordinates fell within designated boundaries, confirming the alignment between LSOA codes and names, and cross-referencing the police force reporting the crime with the corresponding LSOA name of the city to which the police force is affiliated. This assessment helps us understand the limitations of the data and the potential impact on our findings.

### 4.4.1 Preserve the original dataset
To safeguard the integrity of the data, a duplicate copy was created, ensuring the original dataframe remains intact and protected against inadvertent modifications. This precaution helps avoid the need to reload the data if something unintended happens during analysis or changes to the dataframe.

In [6]:
combined_data = combined_data_final.copy(deep=True)

In [7]:
combined_data

Crime ID    Month  \
0       7312a0dc35158e77ef2ae6eb6ee1ecc1c1946e20cb8065...  2022-03   
1       be5a15af709d612e9b39b444377be7b2b9afcca350194f...  2022-03   
2       09698779f4bae0791b25e70122738d8872c8e2cda88df0...  2022-03   
3       d7bb4e184301f37f0f8450b50f874e95a40d4d4e54edd1...  2022-03   
4       5734122a4b9e007d93087a44be315c18302b7d4ae17ac4...  2022-03   
...                                                   ...      ...   
814550  9ff77055ed67fe810427e75fb2ff26e352ccbe709365a8...  2024-03   
814551  36ea75a7e3fe5853606fc50c29295d8fdac4e66cf5711b...  2024-03   
814552  80775f3dd3d42b60ffacb6094893ce3272df153822632c...  2024-03   
814553  7e61ef1ea7dbd0e809d2067f1d6bef9d557a9eb65a98be...  2024-03   
814554  592ea60f980cbcca058d9f9079e41a827d5cf91136da83...  2024-03   

                        Reported by                 Falls within  Longitude  \
0              Thames Valley Police         Thames Valley Police  -0.972323   
1              Thames Valley Police         Thames Valley Police  -0.972323   
2              Thames Valley Police         Thames Valley Police  -0.970516   
3              Thames Valley Police         Thames Valley Police  -0.970516   
4              Thames Valley Police         Thames Valley Police  -0.996750   
...                             ...                          ...        ...   
814550  Cambridgeshire Constabulary  Cambridgeshire Constabulary        NaN   
814551  Cambridgeshire Constabulary  Cambridgeshire Constabulary        NaN   
814552  Cambridgeshire Constabulary  Cambridgeshire Constabulary        NaN   
814553  Cambridgeshire Constabulary  Cambridgeshire Constabulary        NaN   
814554  Cambridgeshire Constabulary  Cambridgeshire Constabulary        NaN   

         Latitude                Location  LSOA code            LSOA name  \
0       51.991460  On or near Osprey Walk  E01017648  Aylesbury Vale 001A   
1       51.991460  On or near Osprey Walk  E01017648  Aylesbury Vale 001A   
2       51.992128  On or near Lime Avenue  E01017648  Aylesbury Vale 001A   
3       51.992128  On or near Lime Avenue  E01017648  Aylesbury Vale 001A   
4       51.997191   On or near Westfields  E01017649  Aylesbury Vale 001B   
...           ...                     ...        ...                  ...   
814550        NaN             No Location        NaN                  NaN   
814551        NaN             No Location        NaN                  NaN   
814552        NaN             No Location        NaN                  NaN   
814553        NaN             No Location        NaN                  NaN   
814554        NaN             No Location        NaN                  NaN   

                          Crime type  \
0       Violence and sexual offences   
1       Violence and sexual offences   
2       Violence and sexual offences   
3       Violence and sexual offences   
4          Criminal damage and arson   
...                              ...   
814550                   Other crime   
814551                   Other crime   
814552                   Other crime   
814553                   Other crime   
814554                   Other crime   

                             Last outcome category  Context  
0       Action to be taken by another organisation      NaN  
1       Action to be taken by another organisation      NaN  
2       Action to be taken by another organisation      NaN  
3       Action to be taken by another organisation      NaN  
4                      Unable to prosecute suspect      NaN  
...                                            ...      ...  
814550                 Unable to prosecute suspect      NaN  
814551                         Under investigation      NaN  
814552                         Under investigation      NaN  
814553                 Unable to prosecute suspect      NaN  
814554                         Under investigation      NaN  

[814555 rows x 12 columns]

### 4.4.2 Remove 'Context' column

In [8]:
# 2. Remove 'Context' column that was filled with 'null' values
combined_data = combined_data.drop('Context', axis=1)
combined_data.columns # displays the name of all columns. 'Context' does not appear anymore

Index(['Crime ID', 'Month', 'Reported by', 'Falls within', 'Longitude',
       'Latitude', 'Location', 'LSOA code', 'LSOA name', 'Crime type',
       'Last outcome category'],
      dtype='object')

### 4.4.3 Check uniqueness of the crimes

In [9]:
#3. replace null crime Id to 'Unknown'
# Calculate the percentage of 'null' values
print('The percentage of null Crime ID values is: %f %%' % (combined_data['Crime ID'].isnull().sum() / len(combined_data['Crime ID']) * 100))

#combined_data['Crime ID'] = combined_data['Crime ID'].fillna('unknown')

The percentage of null Crime ID values is: 11.919269 %


Taking into consideration that almost 12% is a high percentage of data to remove, the 'null' values will be replaced by 'Unknown'.

In [10]:
# replace 'null' values in 'Crime ID' for 'Unknown'
combined_data['Crime ID'].fillna('Unknown', inplace=True)
display(combined_data.isnull().sum()) # shows all the null values in the dataframe, 'Crime ID' has 0

Crime ID                     0
Month                        0
Reported by                  0
Falls within                 0
Longitude                33109
Latitude                 33109
Location                     0
LSOA code                33110
LSOA name                33110
Crime type                   0
Last outcome category    97089
dtype: int64

Due to the fact that a large percentage of 'Crime ID' values are 'Unknown', the unique identifier for each row used will be a combination of the 'Crime ID', 'Month', Latitude', 'Longitude' and 'Crime type' as the same crime can't be comited in the same place at the same time. To have a more accurate representation of it it would be necessary to have a day and time the incident occur.
Considering that 'Crime ID' is an unique value that identifies each crime, to ensure there is no replicated data, this column will be checked for uniqueness and removed all duplicated values.

In [11]:
# check uniqueness of the crimes 
print(len(combined_data), combined_data['Crime ID'].nunique()) 
# if these numbers are equal, there are no repeated crime Ids - this is not the case

# remove duplicates
combined_data.drop_duplicates(subset=["Crime ID", "Latitude", "Longitude", "Month", "Crime type"], inplace=True)

# recheck
print(len(combined_data)) 

814555 716048
790328


It possible to see that out of 814 555 rows, 716 048 are unique 'Crime ID' values, which shows that this variable can't be used as a primary key. Therefore, the combination mentioned previously results in  a total of 790 328 unique values.

In [12]:
display(combined_data.isnull().sum())

Crime ID                     0
Month                        0
Reported by                  0
Falls within                 0
Longitude                32269
Latitude                 32269
Location                     0
LSOA code                32270
LSOA name                32270
Crime type                   0
Last outcome category    73131
dtype: int64

### 4.4.4 Latitude and Longitude analysis
The longitude and latitude values will be heavily used for the geographical analysis of these crimes, it won't be possible to use any rows with eaither column containing 'null' values. These will be deleted.

In [13]:
# 4. remove all nulls from 'Latitude' and 'Longitude'
combined_data.dropna(subset=["Latitude", "Longitude"], inplace=True)
display(combined_data.isnull().sum()) 

Crime ID                     0
Month                        0
Reported by                  0
Falls within                 0
Longitude                    0
Latitude                     0
Location                     0
LSOA code                    1
LSOA name                    1
Crime type                   0
Last outcome category    73110
dtype: int64

To ensure the accuracy of latitude and longitude coordinates within each city, an analysis of coordinate values was conducted for Oxford, Cambridge, and Bristol. Given the need to repeat this process for all three cities, a function was devised to compare boundaries, allowing for the utilization of distinct boundaries for each city.

For precise geographic delineation, bounding box coordinates were employed for Oxford, Cambridge, and Bristol. These coordinates serve as reference points to filter your DataFrame, enabling the isolation of data specific to each city based on geographical regions.The coordinates were determined manually using Google maps.

In [14]:
# Function to filter DataFrame based on city boundaries
def filter_by_bounds(df, bounds):
    """
    Filters the input DataFrame based on latitude and longitude boundaries.

    Parameters:
    df (pandas.DataFrame): The input DataFrame containing 'Latitude' and 'Longitude' columns.
    bounds (dict): A dictionary with the following keys:
        - 'min_lat': Minimum latitude boundary
        - 'max_lat': Maximum latitude boundary
        - 'min_lon': Minimum longitude boundary
        - 'max_lon': Maximum longitude boundary

    Returns:
    pandas.DataFrame: A DataFrame containing only the rows where 'Latitude' and 'Longitude'
                      are within the specified bounds.
    """
    return df[
        (df['Latitude'] >= bounds['min_lat']) &  # Check if Latitude is greater than or equal to min_lat
        (df['Latitude'] <= bounds['max_lat']) &  # Check if Latitude is less than or equal to max_lat
        (df['Longitude'] >= bounds['min_lon']) &  # Check if Longitude is greater than or equal to min_lon
        (df['Longitude'] <= bounds['max_lon'])    # Check if Longitude is less than or equal to max_lon
    ]


# Filter DataFrame for each city based on the presence of city name in 'LSOA name'
oxford_data = combined_data[combined_data['LSOA name'].str.contains('Oxford', case=False, na=False)]
cambridge_data = combined_data[combined_data['LSOA name'].str.contains('Cambridge', case=False, na=False)]
bristol_data = combined_data[combined_data['LSOA name'].str.contains('Bristol', case=False, na=False)]

print('Before checking if the coordinates are correct: Oxford = %d rows; Cambridge = %d rows; Bristol: %d' % (len(oxford_data), len(cambridge_data), len(bristol_data)))

# These boundaries are defined as dictionaries with keys for min and max latitude and longitude
oxford_bounds = {'min_lat': 51.70, 'max_lat': 51.81, 'min_lon': -1.32, 'max_lon': -1.17}
cambridge_bounds = {'min_lat': 52.15, 'max_lat': 52.24, 'min_lon': 0.04, 'max_lon': 0.20}
bristol_bounds = {'min_lat': 51.38, 'max_lat': 51.55, 'min_lon': -2.74, 'max_lon': -2.44}

# Apply the filter_by_bounds function to combined_data for each city's boundaries
oxford_data = filter_by_bounds(oxford_data, oxford_bounds)  # Filter data for Oxford
cambridge_data = filter_by_bounds(cambridge_data, cambridge_bounds)  # Filter data for Cambridge
bristol_data = filter_by_bounds(bristol_data, bristol_bounds)  # Filter data for Bristol

print('After checking if the coordinates are correct: Oxford = %d rows; Cambridge = %d rows; Bristol: %d' % (len(oxford_data), len(cambridge_data), len(bristol_data)))

Before checking if the coordinates are correct: Oxford = 56971 rows; Cambridge = 52794 rows; Bristol: 103710
After checking if the coordinates are correct: Oxford = 31996 rows; Cambridge = 31758 rows; Bristol: 103710


According to the filtering done, its possible to see that all coordinates are within the cities in study.

### 4.4.5 LSOA codes and names analysis
Likewise, the LSOA code and name will be necessary for the Exploratory Data Analysis.

In [15]:
# Remove nulls from 'LSOA code' column
combined_data.dropna(subset=["LSOA code"], inplace=True)


Another preprocessing technique used, to ensure the LSOA codes match according to the LSOA names, and external csv file([reference](https://www.data.gov.uk/dataset/6cf96809-29d1-4b87-8f1b-2b06490ae51c/lower-layer-super-output-areas-december-2021-names-and-codes-in-england-and-wales-v2)) was used to compare these values.

In [16]:
# Load the LSOA reference DataFrame
lsoa_reference = pd.read_csv('LSOA_codesAndNames.csv')

# Print the number of rows in the dataset before analysis
print('Number of rows in the dataset before analyzing if LSOA codes match the names: %d' % len(combined_data))

# Check which LSOA codes in the main DataFrame are not in the reference DataFrame
non_matching_codes = combined_data[~combined_data['LSOA code'].isin(lsoa_reference['LSOA21CD'])]

# Calculate the number of unique non-matching LSOA codes
unique_non_matching_codes = non_matching_codes['LSOA code'].nunique()

# Calculate the total number of LSOA codes in the main DataFrame
total_lsoa_codes = len(combined_data['LSOA code'])

# Calculate the percentage of LSOA codes that don't match
percentage_codes_not_matching = (unique_non_matching_codes / total_lsoa_codes) * 100
print('Percentage of LSOA codes that don\'t match: %.2f%%' % percentage_codes_not_matching)

# Check if corresponding LSOA names for non-matching codes don't match with the reference DataFrame
non_matching_names = non_matching_codes[~non_matching_codes['LSOA name'].isin(lsoa_reference['LSOA21NM'])]

# Calculate the number of unique non-matching LSOA names
unique_non_matching_names = non_matching_names['LSOA name'].nunique()

# Calculate the total number of LSOA names in the main DataFrame
total_lsoa_names = len(combined_data['LSOA name'])

# Calculate the percentage of LSOA names that don't match
percentage_names_not_matching = (unique_non_matching_names / total_lsoa_names) * 100
print('Percentage of LSOA names that don\'t match: %.2f%%' % percentage_names_not_matching)


Number of rows in the dataset before analyzing if LSOA codes match the names: 758058
Percentage of LSOA codes that don't match: 0.02%
Percentage of LSOA names that don't match: 0.02%


After checking a reference file for the existing LSOA codes, it's possible to see that there is a % of LSOA codes and names that are not present in the reference file. This might be because the reference file is not properly updated. To check this, the LSOA code values non matching to the reference were checked for uniqueness in match to the LSOA name. In other words, if an LSOA code is not present in the reference file, is it uniquelly matching a LSOA name in all rows.

In [17]:
# Group non-matching codes by LSOA code and check if each group has only one unique LSOA name
same_lsoa_name = non_matching_codes.groupby('LSOA code')['LSOA name'].nunique().eq(1).all()

# Print result
if same_lsoa_name:
    print('All non-matching LSOA codes correspond to the same LSOA name.')
else:
    print('Non-matching LSOA codes do not correspond to the same LSOA name for all codes.')

All non-matching LSOA codes correspond to the same LSOA name.


This might be because the reference file is not updated and not because the LSOA codes and names are not accurate. So the values are still kept for analysis.

### 4.4.6 'Last outcome category' column
On the other hand, 'Last outcome category' won't be used for the analysis so it's unnecessary to remove null values.

In [18]:
# Delete 'Last outcome category' column
combined_data.drop(columns=["Last outcome category"], inplace=True)

display(combined_data.isnull().sum()) 

Crime ID        0
Month           0
Reported by     0
Falls within    0
Longitude       0
Latitude        0
Location        0
LSOA code       0
LSOA name       0
Crime type      0
dtype: int64

### 4.4.7 'Reported by' and 'Falls within' analysis
It was also noticed that the columns 'Reported by' and 'Falls within' may be equal. To avoid repeating information, if these 2 columns are equal, the column 'Falls within' will be deleted.

In [19]:
# 5. Are the columns 'Reported by' and 'Falls within' equal?
# if true, remove 'Falls within'
if (combined_data['Reported by'] == combined_data['Falls within']).all() == True:
    combined_data = combined_data.drop('Falls within', axis=1)
    print('Column was deleted')
else:
    print('Column was kept')
# Output value is true

Column was deleted


### 4.4.8 'Location' column cleaning
In addition, the column 'Location' appear to have a lot of 'On or near ' as the begining part of the string. To reduce the string content, if all row values have this, it will be deleted from the text value and inserted as title for the location column.

In [20]:
#6. Change location column
#Remove 'On or near' if all rows have it
if combined_data['Location'].str.contains('On or near').all() == True: # checks if all values in the 'Location' column contain the substring 'On or near'
    combined_data['Location'] = combined_data['Location'].str.replace('On or near ', '', regex=False) # This effectively removes the 'On or near' prefix from each location
    combined_data.rename(columns={'Location': 'Location (on/ near)'}, inplace=True) # Rename the column
    print('Changed') #is executed if the 'Location' column was modified 
else:
    print('Unchanged')

Changed


### 4.4.9 Retrieving crimes for Cambridge, Oxford and Bristol
Since the regions being studied are Cambridge, Oxford and Bristol, to ensure only the data related to these cities is used, the rows were filtered by LSOA name containing the names of the cities.

In [21]:
# Filter LSOA codes for the regions being investigated (Cambridge, Bristol and Oxford)
print(combined_data['LSOA name'].unique()) # shows the dataset before comparing the LSOA codes

combined_data = combined_data[combined_data['LSOA name'].str.contains(r'\b(Bristol|Cambridge|Oxford)\b', case=False, na=False)] 
# filters the DataFrame to keep only the rows where the 'LSOA name' contains any of the words 'Bristol', 'Cambridge', or 'Oxford'
# case=False makes the search case-insensitive
# na=False ensures that NaN values are treated as not matching the condition, preventing errors during filtering

print(combined_data['LSOA name'].unique()) # shows the dataset after removing the unnecessary LSOA codes

['Aylesbury Vale 001A' 'Aylesbury Vale 001B' 'Aylesbury Vale 001C' ...
 'South Holland 005D' 'Spelthorne 004F' 'West Suffolk 003D']


C:\Users\MartaMoreira\AppData\Local\Temp\ipykernel_24196\3232922605.py:4: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.
  combined_data = combined_data[combined_data['LSOA name'].str.contains(r'\b(Bristol|Cambridge|Oxford)\b', case=False, na=False)]


['Oxford 001A' 'Oxford 001B' 'Oxford 001C' 'Oxford 001D' 'Oxford 002C'
 'Oxford 002D' 'Oxford 002F' 'Oxford 002G' 'Oxford 003A' 'Oxford 003B'
 'Oxford 003C' 'Oxford 004A' 'Oxford 004B' 'Oxford 004C' 'Oxford 004D'
 'Oxford 005A' 'Oxford 005B' 'Oxford 005C' 'Oxford 005D' 'Oxford 006A'
 'Oxford 006B' 'Oxford 006C' 'Oxford 006D' 'Oxford 006E' 'Oxford 006F'
 'Oxford 007A' 'Oxford 007B' 'Oxford 007C' 'Oxford 007D' 'Oxford 007E'
 'Oxford 008A' 'Oxford 008B' 'Oxford 008E' 'Oxford 008G' 'Oxford 009A'
 'Oxford 009B' 'Oxford 009C' 'Oxford 009D' 'Oxford 010A' 'Oxford 010B'
 'Oxford 010C' 'Oxford 010D' 'Oxford 011A' 'Oxford 011B' 'Oxford 011C'
 'Oxford 011D' 'Oxford 011E' 'Oxford 011F' 'Oxford 011G' 'Oxford 012A'
 'Oxford 012B' 'Oxford 012C' 'Oxford 012D' 'Oxford 013A' 'Oxford 013B'
 'Oxford 013C' 'Oxford 013D' 'Oxford 013E' 'Oxford 013F' 'Oxford 013G'
 'Oxford 014A' 'Oxford 014B' 'Oxford 014C' 'Oxford 015A' 'Oxford 015B'
 'Oxford 015C' 'Oxford 015D' 'Oxford 015E' 'Oxford 015F' 'Oxford 016A'
 'Oxfo

### 4.4.10 Reset index
When performing operations like filtering, merging, or grouping on a DataFrame, the index may become unordered or contain gaps. Resetting the index ensures that the index is reset to a consecutive range starting from 0, without preserving the old index values.

In [22]:
# reset index
combined_data = combined_data.reset_index(drop=True)
display(combined_data)

Crime ID    Month  \
0                                                 Unknown  2022-03   
1                                                 Unknown  2022-03   
2       2dfb0a74a8356c4873be4ca8ca0e72ce3727052b15f906...  2022-03   
3       5d3450347523b0854ace3e26cb5d0a4d00d315c214800f...  2022-03   
4       7f3cdbfc2f36b7e5f5c289e4886e42a4d759203f4a5ccc...  2022-03   
...                                                   ...      ...   
164056  f041cc8af9723794098b1f0405a189eceebf2d96c52390...  2024-03   
164057  7386d2d767baf6c924fb5a0132db7a15ea2e100176af97...  2024-03   
164058  85d61544aa66af647d9d9eac34736d16978e8fdfbf8bf9...  2024-03   
164059  89cd70f2fab906c19d210d06978edde38d92b3262eb198...  2024-03   
164060  07e364b9b3123d27ab453f22f532a9c20d2c9d7feae419...  2024-03   

                        Reported by  Longitude   Latitude  \
0              Thames Valley Police  -1.276727  51.793600   
1              Thames Valley Police  -1.276321  51.789974   
2              Thames Valley Police  -1.273320  51.794505   
3              Thames Valley Police  -1.269903  51.790563   
4              Thames Valley Police  -1.275768  51.789188   
...                             ...        ...        ...   
164056  Cambridgeshire Constabulary   0.119019  52.169742   
164057  Cambridgeshire Constabulary   0.124102  52.172052   
164058  Cambridgeshire Constabulary   0.108288  52.165457   
164059  Cambridgeshire Constabulary   0.116195  52.167851   
164060  Cambridgeshire Constabulary   0.107727  52.166285   

           Location (on/ near)  LSOA code       LSOA name  \
0       Sports/Recreation Area  E01028594     Oxford 001A   
1              Five Mile Drive  E01028594     Oxford 001A   
2             Jordan Hill Road  E01028594     Oxford 001A   
3                Lovelace Road  E01028594     Oxford 001A   
4              Rothafield Road  E01028594     Oxford 001A   
...                        ...        ...             ...   
164056               Royal Way  E01035529  Cambridge 015F   
164057        Pinnington Close  E01035529  Cambridge 015F   
164058                Elm Road  E01035530  Cambridge 015G   
164059            Exeter Close  E01035530  Cambridge 015G   
164060            Hauxton Road  E01035530  Cambridge 015G   

                          Crime type  
0              Anti-social behaviour  
1              Anti-social behaviour  
2                              Drugs  
3                              Drugs  
4                              Drugs  
...                              ...  
164056  Violence and sexual offences  
164057  Violence and sexual offences  
164058                  Public order  
164059  Violence and sexual offences  
164060  Violence and sexual offences  

[164061 rows x 9 columns]

### 4.4.11 Saving pre-processed data to a new csv file

Saving pre-processed data into a new CSV file serves several purposes essential for data management and analysis. Firstly, it ensures that the pre-processing steps, such as cleaning, filtering, or feature engineering, are preserved and can be easily replicated in future analyses. By saving the pre-processed data, researchers and analysts can maintain a clear record of the transformations applied to the original dataset, aiding in transparency and reproducibility of results. Moreover, it facilitates data sharing and collaboration, allowing others to access and utilize the cleaned data for their own research purposes. Additionally, saving pre-processed data into a separate CSV file helps streamline data management practices by separating raw data from processed data, reducing clutter and potential confusion when working with multiple datasets. Finally, having a clean and well-organized CSV file enables efficient data loading and manipulation in various analytical tools and programming environments, enhancing the overall workflow efficiency. Overall, saving pre-processed data into a new CSV file is a crucial step in the data analysis pipeline, promoting transparency, reproducibility, and effective data management practices.

In [23]:
# save it into a csv file
combined_data.to_csv('Combined_data_EDA.csv', index=False)

### 4.4.12 Review pre-processed data
The pre-processed data can be seen in the following table.

In [24]:
display(combined_data.head(5))

Crime ID    Month  \
0                                            Unknown  2022-03   
1                                            Unknown  2022-03   
2  2dfb0a74a8356c4873be4ca8ca0e72ce3727052b15f906...  2022-03   
3  5d3450347523b0854ace3e26cb5d0a4d00d315c214800f...  2022-03   
4  7f3cdbfc2f36b7e5f5c289e4886e42a4d759203f4a5ccc...  2022-03   

            Reported by  Longitude   Latitude     Location (on/ near)  \
0  Thames Valley Police  -1.276727  51.793600  Sports/Recreation Area   
1  Thames Valley Police  -1.276321  51.789974         Five Mile Drive   
2  Thames Valley Police  -1.273320  51.794505        Jordan Hill Road   
3  Thames Valley Police  -1.269903  51.790563           Lovelace Road   
4  Thames Valley Police  -1.275768  51.789188         Rothafield Road   

   LSOA code    LSOA name             Crime type  
0  E01028594  Oxford 001A  Anti-social behaviour  
1  E01028594  Oxford 001A  Anti-social behaviour  
2  E01028594  Oxford 001A                  Drugs  
3  E01028594  Oxford 001A                  Drugs  
4  E01028594  Oxford 001A                  Drugs

## 4.7 Workflow Visualization

Visualizing the preprocessing workflow provides a clear understanding of the steps involved and their sequence. In this section, we present a diagram or flowchart depicting the entire preprocessing process, from data collection to quality assessment. This visualization aids in comprehending the workflow and identifying areas for improvement.

the Flowchart can be acessed through: [Flow Chart](https://lucid.app/lucidchart/2733135c-9f30-4499-aae7-dee08df51395/edit?viewport_loc=-1293%2C-587%2C5293%2C2712%2C0_0&invitationId=inv_550db952-4952-4115-9bd8-180489ab2376
) 

In [25]:
combined_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 164061 entries, 0 to 164060
Data columns (total 9 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   Crime ID             164061 non-null  object 
 1   Month                164061 non-null  object 
 2   Reported by          164061 non-null  object 
 3   Longitude            164061 non-null  float64
 4   Latitude             164061 non-null  float64
 5   Location (on/ near)  164061 non-null  object 
 6   LSOA code            164061 non-null  object 
 7   LSOA name            164061 non-null  object 
 8   Crime type           164061 non-null  object 
dtypes: float64(2), object(7)
memory usage: 11.3+ MB
